[BUG] Panic: `interface conversion: interface {} is int, not float64` during template parsing

```

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v3.7.1

		projectdiscovery.io

panic: interface conversion: interface {} is int, not float64 [recovered, repanicked]

goroutine 2515 [running]:
gopkg.in/yaml%2ev2.handleErr(0x34ee31e99c20)
	gopkg.in/yaml.v2@v2.4.0/yaml.go:249 +0x65
panic({0x344ace0?, 0x34ee2eb00ed0?})
	runtime/panic.go:860 +0x13a
gopkg.in/yaml%2ev2.handleErr(0x34ee31e99828)
	gopkg.in/yaml.v2@v2.4.0/yaml.go:249 +0x65
panic({0x344ace0?, 0x34ee2eb00ed0?})
	runtime/panic.go:860 +0x13a
github.com/projectdiscovery/dsl.init.0.func56({0x34ee2ef7ccd0?, 0x9?, 0x10?})
	github.com/projectdiscovery/dsl@v0.8.14/dsl.go:875 +0xf9
github.com/projectdiscovery/dsl.dslFunction.Exec({0x0, {0x3ac3d54, 0x9}, 0x0, {0x34ee2d983a20, 0x1, 0x1}, 0x474c308}, {0x34ee2ef7ccd0, 0x1, ...})
	github.com/projectdiscovery/dsl@v0.8.14/func.go:57 +0x2b6
github.com/Knetic/govaluate.planFunction.makeFunctionStage.func1({0x34ee2efa9360?, 0x327d680?}, {0x327d680, 0x49009a8}, {0x3ba191d?, 0x200ee31e98b01?})
	github.com/Knetic/govaluate@v3.0.1-0.20171022003610-9aa49832a739+incompatible/evaluationStage.go:245 +0xac
github.com/Knetic/govaluate.EvaluableExpression.evaluateStage({{0x3bb4d9c, 0x22}, 0x1, {0x34ee2cf85b00, 0x9, 0x10}, 0x34ee2efa93b0, {0x34ee2ef807c2, 0x1a}}, 0x34ee2efa93b0, ...)
	github.com/Knetic/govaluate@v3.0.1-0.20171022003610-9aa49832a739+incompatible/EvaluableExpression.go:232 +0x5ad
github.com/Knetic/govaluate.EvaluableExpression.Eval({{0x3bb4d9c, 0x22}, 0x1, {0x34ee2cf85b00, 0x9, 0x10}, 0x34ee2efa93b0, {0x34ee2ef807c2, 0x1a}}, {0x4778aa0, ...})
	github.com/Knetic/govaluate@v3.0.1-0.20171022003610-9aa49832a739+incompatible/EvaluableExpression.go:163 +0x125
github.com/Knetic/govaluate.EvaluableExpression.Evaluate({{0x3bb4d9c, 0x22}, 0x1, {0x34ee2cf85b00, 0x9, 0x10}, 0x34ee2efa93b0, {0x34ee2ef807c2, 0x1a}}, 0x34ee2eb00c90)
	github.com/Knetic/govaluate@v3.0.1-0.20171022003610-9aa49832a739+incompatible/EvaluableExpression.go:137 +0x94
github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/expressions.evaluate({0x34ee2ef807c0, 0x1e}, 0x34ee2eb00c90)
	github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/expressions/expressions.go:66 +0x1e9
github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/expressions.Evaluate(...)
	github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/expressions/expressions.go:31
github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/variables.evaluateVariableValueWithMap({0x34ee2ef807c0, 0x1e}, 0x34ee2eb06e90?)
	github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/variables/variables.go:147 +0x1e
github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/variables.(*Variable).Evaluate.func1(...)
	github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/variables/variables.go:87
github.com/projectdiscovery/nuclei/v3/pkg/utils.(*InsertionOrderedStringMap).ForEach(...)
	github.com/projectdiscovery/nuclei/v3/pkg/utils/insertion_ordered_map.go:107
github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/variables.(*Variable).Evaluate(0x34ee2d4b4780, 0x34ee31e99220)
	github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/variables/variables.go:76 +0x35d
github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/variables.(*Variable).UnmarshalYAML(0x34ee2d4b4780, 0x70381bc397a0?)
	github.com/projectdiscovery/nuclei/v3/pkg/protocols/common/variables/variables.go:49 +0x2c5
gopkg.in/yaml%2ev2.(*decoder).callUnmarshaler(0x34ee2ef7a4e0, 0x34ee2efa33b0, {0x70381bc397a0, 0x34ee2d4b4780})
	gopkg.in/yaml.v2@v2.4.0/decode.go:270 +0x94
gopkg.in/yaml%2ev2.(*decoder).prepare(0x34ee2ef7a4e0, 0x34ee2efa33b0, {0x36d9c80?, 0x34ee2d4b4780?, 0x34ee2eb3fac0?})
	gopkg.in/yaml.v2@v2.4.0/decode.go:313 +0x256
gopkg.in/yaml%2ev2.(*decoder).unmarshal(0x34ee2ef7a4e0, 0x34ee2efa33b0, {0x36d9c80?, 0x34ee2d4b4780?, 0x0?})
	gopkg.in/yaml.v2@v2.4.0/decode.go:364 +0x17e
gopkg.in/yaml%2ev2.(*decoder).mappingStruct(0x34ee2ef7a4e0, 0x34ee2efa2070, {0x39eac80?, 0x34ee2d4b4588?, 0x34ee2ef9f6d8?})
	gopkg.in/yaml.v2@v2.4.0/decode.go:767 +0xe2b
gopkg.in/yaml%2ev2.(*decoder).mapping(0x34ee2ef7a4e0?, 0x34ee2efa2070?, {0x39eac80?, 0x34ee2d4b4588?, 0x34ee2d542008?})
	gopkg.in/yaml.v2@v2.4.0/decode.go:626 +0x465
gopkg.in/yaml%2ev2.(*decoder).unmarshal(0x34ee2ef7a4e0, 0x34ee2efa2070, {0x319f120?, 0x34ee2d4b4588?, 0x2c0?})
	gopkg.in/yaml.v2@v2.4.0/decode.go:372 +0x1e5
gopkg.in/yaml%2ev2.(*decoder).callUnmarshaler.func1({0x319f120?, 0x34ee2d4b4588?})
	gopkg.in/yaml.v2@v2.4.0/decode.go:272 +0xe5
github.com/projectdiscovery/nuclei/v3/pkg/templates.(*Template).UnmarshalYAML(0x34ee2d4b42c8, 0x34ee2eb3f8e0)
	github.com/projectdiscovery/nuclei/v3/pkg/templates/templates.go:339 +0x58
gopkg.in/yaml%2ev2.(*decoder).callUnmarshaler(0x34ee2ef7a4e0, 0x34ee2efa2070, {0x70381bc396a0, 0x34ee2d4b42c8})
	gopkg.in/yaml.v2@v2.4.0/decode.go:270 +0x94
gopkg.in/yaml%2ev2.(*decoder).prepare(0x34ee2ef7a4e0, 0x34ee2efa2070, {0x39ea960?, 0x34ee2d4b42c8?, 0xdc5f86?})
	gopkg.in/yaml.v2@v2.4.0/decode.go:313 +0x256
gopkg.in/yaml%2ev2.(*decoder).unmarshal(0x34ee2ef7a4e0, 0x34ee2efa2070, {0x39ea960?, 0x34ee2d4b42c8?, 0x16?})
	gopkg.in/yaml.v2@v2.4.0/decode.go:364 +0x17e
gopkg.in/yaml%2ev2.(*decoder).document(...)
	gopkg.in/yaml.v2@v2.4.0/decode.go:384
gopkg.in/yaml%2ev2.(*decoder).unmarshal(0x34ee2ef7a4e0, 0x34ee2efa2000, {0x39ea960?, 0x34ee2d4b42c8?, 0x10034ee2d760000?})
	gopkg.in/yaml.v2@v2.4.0/decode.go:360 +0x15d
gopkg.in/yaml%2ev2.unmarshal({0x34ee2d762400, 0xb72, 0xc00}, {0x39d23e0, 0x34ee2d4b42c8}, 0x1)
	gopkg.in/yaml.v2@v2.4.0/yaml.go:148 +0x37c
gopkg.in/yaml%2ev2.UnmarshalStrict(...)
	gopkg.in/yaml.v2@v2.4.0/yaml.go:89
github.com/projectdiscovery/nuclei/v3/pkg/templates.(*Parser).ParseTemplate(0x34ee2d4d3600, {0x34ee2ef841e0, 0x24}, {0x47a05b0, 0x34ee2d4d3620})
	github.com/projectdiscovery/nuclei/v3/pkg/templates/parser.go:173 +0x266
github.com/projectdiscovery/nuclei/v3/pkg/templates.(*Parser).LoadTemplate(0x24?, {0x34ee2ef841e0, 0x24}, {0x339bb20?, 0x34ee2ce78900}, {0x0, 0x0, 0x0}, {0x47a05b0, 0x34ee2d4d3620})
	github.com/projectdiscovery/nuclei/v3/pkg/templates/parser.go:91 +0x8f
github.com/projectdiscovery/nuclei/v3/pkg/catalog/loader.(*Store).LoadTemplatesWithTags.func2({0x34ee2ef841e0, 0x24})
	github.com/projectdiscovery/nuclei/v3/pkg/catalog/loader/loader.go:833 +0x17c
created by github.com/projectdiscovery/nuclei/v3/pkg/catalog/loader.(*Store).LoadTemplatesWithTags in goroutine 1
	github.com/projectdiscovery/nuclei/v3/pkg/catalog/loader/loader.go:811 +0x4f1
```

--- a/dsl.go
+++ b/dsl.go
@@ -333,8 +333,9 @@ func init() {
 			readLimit := DefaultMaxDecompressionSize
 
 			if len(args) > 1 {
-				if limit, ok := args[1].(float64); ok {
-					readLimit = int64(limit)
+				limit, err := toInt64(args[1])
+				if err == nil {
+					readLimit = limit
 				}
 			}
 
@@ -377,8 +378,9 @@ func init() {
 			readLimit := DefaultMaxDecompressionSize
 
 			if len(args) > 1 {
-				if limit, ok := args[1].(float64); ok {
-					readLimit = int64(limit)
+				limit, err := toInt64(args[1])
+				if err == nil {
+					readLimit = limit
 				}
 			}
 
@@ -425,8 +427,9 @@ func init() {
 			readLimit := DefaultMaxDecompressionSize
 
 			if len(args) > 1 {
-				if limit, ok := args[1].(float64); ok {
-					readLimit = int64(limit)
+				limit, err := toInt64(args[1])
+				if err == nil {
+					readLimit = limit
 				}
 			}
 
@@ -864,7 +867,6 @@ func init() {
 		"(length uint, optionalCharSet string) string",
 		false,
 		func(args ...interface{}) (interface{}, error) {
-			var length int
 			charSet := letters + numbers
 
 			argSize := len(args)
@@ -872,7 +874,10 @@ func init() {
 				return nil, ErrInvalidDslFunction
 			}
 
-			length = int(args[0].(float64))
+			length, err := toInt(args[0])
+			if err != nil {
+				return nil, err
+			}
 
 			if argSize == 2 {
 				inputCharSet := toString(args[1])
@@ -886,7 +891,6 @@ func init() {
 		"(length uint, optionalBadChars string) string",
 		false,
 		func(args ...interface{}) (interface{}, error) {
-			length := 0
 			badChars := ""
 
 			argSize := len(args)
@@ -894,7 +898,10 @@ func init() {
 				return nil, ErrInvalidDslFunction
 			}
 
-			length = int(args[0].(float64))
+			length, err := toInt(args[0])
+			if err != nil {
+				return nil, err
+			}
 
 			if argSize == 2 {
 				badChars = toString(args[1])
@@ -906,7 +913,6 @@ func init() {
 		"(length uint, optionalBadChars string) string",
 		false,
 		func(args ...interface{}) (interface{}, error) {
-			var length int
 			badChars := ""
 
 			argSize := len(args)
@@ -914,7 +920,10 @@ func init() {
 				return nil, ErrInvalidDslFunction
 			}
 
-			length = int(args[0].(float64))
+			length, err := toInt(args[0])
+			if err != nil {
+				return nil, err
+			}
 
 			if argSize == 2 {
 				badChars = toString(args[1])
@@ -931,7 +940,10 @@ func init() {
 				return nil, ErrInvalidDslFunction
 			}
 
-			length := int(args[0].(float64))
+			length, err := toInt(args[0])
+			if err != nil {
+				return nil, err
+			}
 			badNumbers := ""
 
 			if argSize == 2 {
@@ -954,10 +966,18 @@ func init() {
 			max := math.MaxInt32
 
 			if argSize >= 1 {
-				min = int(args[0].(float64))
+				convertedMin, err := toInt(args[0])
+				if err != nil {
+					return nil, err
+				}
+				min = convertedMin
 			}
 			if argSize == 2 {
-				max = int(args[1].(float64))
+				convertedMax, err := toInt(args[1])
+				if err != nil {
+					return nil, err
+				}
+				max = convertedMax
 			}
 
 			rint, err := randint.IntN(max - min)
@@ -1001,7 +1021,11 @@ func init() {
 			if argSize != 0 && argSize != 1 {
 				return nil, ErrInvalidDslFunction
 			} else if argSize == 1 {
-				seconds = int(args[0].(float64))
+				convertedSeconds, err := toInt(args[0])
+				if err != nil {
+					return nil, err
+				}
+				seconds = convertedSeconds
 			}
 
 			offset := time.Now().Add(time.Duration(seconds) * time.Second)
@@ -1044,7 +1068,10 @@ func init() {
 			if len(args) != 1 {
 				return nil, ErrInvalidDslFunction
 			}
-			seconds := args[0].(float64)
+			seconds, err := toInt(args[0])
+			if err != nil {
+				return nil, err
+			}
 			time.Sleep(time.Duration(seconds) * time.Second)
 			return true, nil
 		}))
@@ -1153,8 +1180,9 @@ func init() {
 		return toBool(args[0]), nil
 	}))
 	MustAddFunction(NewWithPositionalArgs("dec_to_hex", 1, true, func(args ...interface{}) (interface{}, error) {
-		if number, ok := args[0].(float64); ok {
-			hexNum := strconv.FormatInt(int64(number), 16)
+		number, err := toInt64(args[0])
+		if err == nil {
+			hexNum := strconv.FormatInt(number, 16)
 			return toString(hexNum), nil
 		}
 		return nil, fmt.Errorf("invalid number: %T", args[0])

--- a/dsl_test.go
+++ b/dsl_test.go
@@ -633,6 +633,81 @@ func TestRandDslExpressions(t *testing.T) {
 	}
 }
 
+func TestNestedNumericDslExpressions(t *testing.T) {
+	tests := []struct {
+		expression string
+		assert     func(t *testing.T, result interface{})
+	}{
+		{
+			expression: `rand_base(rand_int(5, 6), "abc")`,
+			assert: func(t *testing.T, result interface{}) {
+				require.Regexp(t, regexp.MustCompile(`^[abc]{5}$`), toString(result))
+			},
+		},
+		{
+			expression: `rand_text_alpha(rand_int(5, 6), "abc")`,
+			assert: func(t *testing.T, result interface{}) {
+				require.Regexp(t, regexp.MustCompile(`^[A-Zd-z]{5}$`), toString(result))
+			},
+		},
+		{
+			expression: `rand_text_alphanumeric(rand_int(5, 6), "ab12")`,
+			assert: func(t *testing.T, result interface{}) {
+				require.Regexp(t, regexp.MustCompile(`^[03-9A-Zc-z]{5}$`), toString(result))
+			},
+		},
+		{
+			expression: `rand_text_numeric(rand_int(5, 6), 123)`,
+			assert: func(t *testing.T, result interface{}) {
+				require.Regexp(t, regexp.MustCompile(`^[0456789]{5}$`), toString(result))
+			},
+		},
+		{
+			expression: `dec_to_hex(rand_int(16, 17))`,
+			assert: func(t *testing.T, result interface{}) {
+				require.Equal(t, "10", result)
+			},
+		},
+		{
+			expression: `date_time("%Y", to_unix_time("2022-01-13"))`,
+			assert: func(t *testing.T, result interface{}) {
+				require.Equal(t, "2022", result)
+			},
+		},
+		{
+			expression: `unix_time(rand_int(1, 2)) > unix_time()`,
+			assert: func(t *testing.T, result interface{}) {
+				require.Equal(t, true, result)
+			},
+		},
+		{
+			expression: `len(gzip_decode(gzip("hello world"), rand_int(1, 2)))`,
+			assert: func(t *testing.T, result interface{}) {
+				require.Equal(t, float64(1), result)
+			},
+		},
+		{
+			expression: `len(zlib_decode(zlib("hello world"), rand_int(1, 2)))`,
+			assert: func(t *testing.T, result interface{}) {
+				require.Equal(t, float64(1), result)
+			},
+		},
+		{
+			expression: `len(inflate(deflate("hello world"), rand_int(1, 2)))`,
+			assert: func(t *testing.T, result interface{}) {
+				require.Equal(t, float64(1), result)
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.expression, func(t *testing.T) {
+			result := evaluateExpression(t, test.expression)
+			test.assert(t, result)
+		})
+	}
+}
+
 func TestFakerDslExpressions(t *testing.T) {
 	zeroArgsFakerFunctions := []string{
 		"rand_ach_account_number",

--- a/go.mod
+++ b/go.mod
@@ -21,6 +21,7 @@ require (
 	github.com/projectdiscovery/utils v0.10.0
 	github.com/sashabaranov/go-openai v1.37.0
 	github.com/spaolacci/murmur3 v1.1.0
+	github.com/spf13/cast v1.10.0
 	github.com/stretchr/testify v1.11.1
 	github.com/vulncheck-oss/go-exploit v1.51.0
 	golang.org/x/exp v0.0.0-20250620022241-b7579e27df2b

--- a/util.go
+++ b/util.go
@@ -15,6 +15,7 @@ import (
 	"github.com/pkg/errors"
 	"github.com/projectdiscovery/utils/html"
 	randint "github.com/projectdiscovery/utils/rand"
+	"github.com/spf13/cast"
 )
 
 const (
@@ -138,6 +139,44 @@ func toBool(data interface{}) bool {
 	}
 }
 
+func numericCastInput(data interface{}) (interface{}, error) {
+	switch value := data.(type) {
+	case int, int8, int16, int32, int64:
+		return value, nil
+	case uint, uint8, uint16, uint32, uint64:
+		return value, nil
+	case float32, float64:
+		return value, nil
+	case string:
+		trimmed := strings.TrimSpace(value)
+		if trimmed == "" {
+			return nil, fmt.Errorf("invalid number: %T", data)
+		}
+		if _, err := strconv.ParseFloat(trimmed, 64); err != nil {
+			return nil, err
+		}
+		return trimmed, nil
+	default:
+		return nil, fmt.Errorf("invalid number: %T", data)
+	}
+}
+
+func toInt(data interface{}) (int, error) {
+	value, err := numericCastInput(data)
+	if err != nil {
+		return 0, err
+	}
+	return cast.ToIntE(value)
+}
+
+func toInt64(data interface{}) (int64, error) {
+	value, err := numericCastInput(data)
+	if err != nil {
+		return 0, err
+	}
+	return cast.ToInt64E(value)
+}
+
 func insertInto(s string, interval int, sep rune) string {
 	var buffer bytes.Buffer
 	before := interval - 1
@@ -206,10 +245,12 @@ func parseTimeOrNow(arguments []interface{}) (time.Time, error) {
 				return time.Time{}, errors.New("invalid argument type")
 			}
 			currentTime = time.Unix(unixTime, 0)
-		case int64, float64:
-			currentTime = time.Unix(int64(inputUnixTime.(float64)), 0)
 		default:
-			return time.Time{}, errors.New("invalid argument type")
+			unixTime, err := toInt64(inputUnixTime)
+			if err != nil {
+				return time.Time{}, errors.New("invalid argument type")
+			}
+			currentTime = time.Unix(unixTime, 0)
 		}
 	} else {
 		currentTime = time.Now()

